python/ray/ml/train/integrations/huggingface/huggingface_trainer.py

```diff
@@ -146,52 +174,391 @@ def trainer_init_per_worker(train_dataset, eval_dataset, **config):
     """
 
     def __init__(
         self,
+        *,
         trainer_init_per_worker: Callable[
-            [TorchDataset, Optional[TorchDataset], Any], Trainer
+            [TorchDataset, Optional[TorchDataset], Any], transformers.trainer.Trainer
         ],
+        datasets: Dict[str, GenDataset],
         trainer_init_config: Optional[Dict] = None,
         torch_config: Optional[TorchConfig] = None,
         scaling_config: Optional[ScalingConfig] = None,
         run_config: Optional[RunConfig] = None,
-        datasets: Optional[Dict[str, GenDataset]] = None,
         preprocessor: Optional[Preprocessor] = None,
         resume_from_checkpoint: Optional[Checkpoint] = None,
     ):
 
-        self._validate_train_loop_per_worker(
+        self._validate_trainer_init_per_worker(
             trainer_init_per_worker, "trainer_init_per_worker"
         )
 
-        assert TRAIN_DATASET_KEY in datasets
-        assert all(
+        if TRAIN_DATASET_KEY not in datasets:
+            raise KeyError(
+                f"'{TRAIN_DATASET_KEY}' key must be preset in `datasets`. "
+                f"Got {list(self.datasets.keys())}"
+            )
+        if not all(
             key in (TRAIN_DATASET_KEY, EVALUATION_DATASET_KEY) for key in datasets
-        )
+        ):
+            raise KeyError(
+                f"Only '{TRAIN_DATASET_KEY}' and '{EVALUATION_DATASET_KEY}' "
+                "keys can be preset in `datasets`. "
+                f"Got {list(self.datasets.keys())}"
+            )
 
         super().__init__(
-            self._create_train_func(trainer_init_per_worker),
-            trainer_init_config,
-            torch_config,
-            scaling_config,
-            run_config,
-            datasets,
-            preprocessor,
-            resume_from_checkpoint,
+            train_loop_per_worker=self._create_train_func(trainer_init_per_worker),
+            train_loop_config=trainer_init_config,
+            torch_config=torch_config,
+            scaling_config=scaling_config,
+            run_config=run_config,
+            datasets=datasets,
+            preprocessor=preprocessor,
+            resume_from_checkpoint=resume_from_checkpoint,
         )
 
-    def _create_train_func(self, trainer_init_per_worker):
+    def __new__(cls, *args, **kwargs):
+        """Store the init args as attributes so this can be merged with Tune hparams."""
+        # This if will be entered in the driver-side Trainer.
+        # The Trainer inside the trainable will have a dict
+        # checkpoint created here.
+        # This is required to ensure that the dir syncing logic
+        # is used instead of serializing several gigabytes of data
+        # when a Checkpoint is sent to a Ray Actor.
+        if (
+            "resume_from_checkpoint" in kwargs
+            and kwargs["resume_from_checkpoint"]._local_path
+        ):
+            checkpoint_path = kwargs["resume_from_checkpoint"].to_directory()
```

Comment:
This will copy the checkpoint directory - do we want that? It seems we're read-only so we should be able to use the current directory. So either we just use `checkpoint_path = kwargs["resume_from_checkpoint"]._local_path` or, preferably `checkpoint_type, checkpoint_path = kwargs["resume_from_checkpoint"].get_internal_representation()`

Reply:
Done, thanks for the explanation!